File: heap_PriorityQueue/kthLargestElementInAnArray.py
import heapq


class Solution:
    def findKthLargestMySolution(self, nums: list[int], k: int) -> int:
        nums = [-x for x in nums]
        heapq.heapify(nums)
        result = []
        while k > 0:
            result.append(heapq.heappop(nums))
            k -= 1

        return -result[-1]
    
    def findKthLargest(self, nums: list[int], k: int) -> int:
        k = len(nums) - k
        def quickSelect(l, r):
            pivot, p = nums[r], l
            for i in range(l, r):
                if nums[i] <= pivot:
                    nums[p], nums[i] = nums[i], nums[p]
                    p += 1
            nums[p], nums[r] = nums[r], nums[p]
            if p > k:
                return quickSelect(l, p - 1)
            elif p < k:
                return quickSelect(p + 1, r)
            else:
                return nums[p]
        
        return quickSelect(0, len(nums) - 1)


if __name__ == "__main__":
    nums = [3, 2, 1, 5, 6, 4]
    k = 2
    print("\"My\" Solution:")
    print(Solution().findKthLargestMySolution(nums, k))
    print("\"Quick Select\" Solution:")
    print(Solution().findKthLargest(nums, k))

File: heap_PriorityQueue/cppRealizations/kthLargestElementInAnArray.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;


class Solution {
public:
    int quickSelect(vector<int>& nums, int left, int right, int k) {
        int pivot = nums[right];
        int p = left;
        for(int i = left; i < right; i++){
            if(nums[i] < pivot){
                swap(nums[i], nums[p]);
                p++;
            }
        }
        swap(nums[p], nums[right]);
        if(p + 1 == k){
            return nums[p];
        }else if(p + 1 < k){
            return quickSelect(nums, p + 1, right, k);
        }else{
            return quickSelect(nums, left, p - 1, k);
        }
    }
    int findKthLargest(vector<int>& nums, int k) {
        k = nums.size() - k + 1;
        return quickSelect(nums, 0, nums.size() - 1, k);
    }
};


int main(){
    vector<int> nums = {3,2,1,5,6,4};
    int k = 2;
    cout << Solution().findKthLargest(nums, k) << endl;
    return 0;
}